src/mongoc/mongoc-stream.c

```diff
@@ -61,6 +61,32 @@ mongoc_stream_close (mongoc_stream_t *stream)
 
 
 /**
+ * mongoc_stream_failed:
+ * @stream: A mongoc_stream_t.
+ *
+ * Frees any resources referenced by @stream, including the memory allocation
+ * for @stream.
+ * This handler is called upon stream failure, such as network errors, invalid replies
+ * or replicaset reconfigures deleteing the stream
+ */
+void
+mongoc_stream_failed (mongoc_stream_t *stream)
+{
+   ENTRY;
+
+   bson_return_if_fail(stream);
+
+   if (stream->failed) {
```

Comment:
Why would it be a buffer?
The top level stream is my provided stream abstraction.

It may be a buffer for normal mongoc behavior, in which case its fine to destroy() it in stream_buffered_destroy() as mongoc doesn't care for the difference